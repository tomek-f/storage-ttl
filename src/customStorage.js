import noop from '@tomekf/noop';

import setJsonWithTtl from './setJsonWithTtl';
import getJsonWithTtl from './getJsonWithTtl';
import getSupport from './getSupport';

const { parse: fromJson, stringify: toJson } = JSON;

export const setJson = storage => (name, obj) => storage.setItem(name, toJson(obj));
export const getJson = storage => name => fromJson(storage.getItem(name));
export const length = storage => () => storage.length;

const setItem = storage => (name, obj) => storage.setItem(name, obj);
const getItem = storage => name => storage.getItem(name);
const removeItem = storage => name => storage.removeItem(name);
const key = storage => int => storage.key(int);
const clear = storage => () => storage.clear();

// no support for *Storage.name, *Storage['name'] support for now
const customStorage = (storage, name) => {
  const support = getSupport(storage);
  const storageLength = support ? length(storage) : noop;

  return {
    get length() {
      return storageLength();
    },
    support,
    name,
    originalStorage: storage,
    setJson: support && storage.setItem ? setJson(storage) : noop,
    getJson: support && storage.getItem ? getJson(storage) : noop,
    setJsonWithTtl: support && storage.setItem ? setJsonWithTtl(storage) : noop,
    getJsonWithTtl: support && storage.getItem && storage.removeItem
      ? getJsonWithTtl(storage) : noop,
    setItem: support && storage.setItem ? setItem(storage) : noop,
    getItem: support && storage.getItem ? getItem(storage) : noop,
    removeItem: support && storage.removeItem ? removeItem(storage) : noop,
    key: support && storage.key ? key(storage) : noop,
    clear: support && storage.clear ? clear(storage) : noop,
  };
};

export default customStorage;
