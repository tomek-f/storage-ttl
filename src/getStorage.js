import hardLog from '@tomekf/hard-log';

export default function getStorage(storageName = 'localStorage') {
  let storage;

  try {
    storage = global[storageName];
  } catch (err) {
    hardLog.warn('@tomekf/storage-tt: no localStorage', err);
  }

  return storage;
}
