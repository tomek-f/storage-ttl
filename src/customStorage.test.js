import '@tomekf/storage-mock/dist/auto';

import customStorage from './customStorage';
import getStorage from './getStorage';

// only snapshots and setJson, getJson and length tested, other are native (mocked) methods and/or tested elsewhere

test('snapshot localStorage', () => {
  expect(customStorage(getStorage(), 'ls++')).toMatchSnapshot();
});

test('snapshot sessionStorage', () => {
  expect(customStorage(getStorage('sessionStorage'), 'ss++')).toMatchSnapshot();
});

test('snapshot BSStorage', () => {
  expect(customStorage(getStorage('BSStorage'), 'BSs++')).toMatchSnapshot();
});

const myStorage = customStorage(getStorage(), 'myStorage');

test('setJson and getJson', () => {
  myStorage.setJson('a', { b: 2, c: 3 });
  expect(myStorage.getJson('a')).toEqual({ b: 2, c: 3 });
});

test('length', () => {
  expect(myStorage.length).toBe(1);
  myStorage.setItem('b', true);
  myStorage.setItem('c', 'fff');
  expect(myStorage.length).toBe(3);
});
