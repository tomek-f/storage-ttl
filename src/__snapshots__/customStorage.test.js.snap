// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot BSStorage 1`] = `
Object {
  "clear": [Function],
  "getItem": [Function],
  "getJson": [Function],
  "getJsonWithTtl": [Function],
  "key": [Function],
  "length": undefined,
  "name": "BSs++",
  "originalStorage": undefined,
  "removeItem": [Function],
  "setItem": [Function],
  "setJson": [Function],
  "setJsonWithTtl": [Function],
  "support": false,
}
`;

exports[`snapshot localStorage 1`] = `
Object {
  "clear": [Function],
  "getItem": [Function],
  "getJson": [Function],
  "getJsonWithTtl": [Function],
  "key": [Function],
  "length": 0,
  "name": "ls++",
  "originalStorage": Storage {},
  "removeItem": [Function],
  "setItem": [Function],
  "setJson": [Function],
  "setJsonWithTtl": [Function],
  "support": true,
}
`;

exports[`snapshot sessionStorage 1`] = `
Object {
  "clear": [Function],
  "getItem": [Function],
  "getJson": [Function],
  "getJsonWithTtl": [Function],
  "key": [Function],
  "length": 0,
  "name": "ss++",
  "originalStorage": Storage {},
  "removeItem": [Function],
  "setItem": [Function],
  "setJson": [Function],
  "setJsonWithTtl": [Function],
  "support": true,
}
`;
